yt/frontends/gadget/io.py

```diff
@@ -77,13 +80,58 @@ def _yield_coordinates(self, data_file, needed_ptype=None):
             yield key, pos
         f.close()
 
+    def _generate_smoothing_length(self, data_files, kdtree):
+        if not self.ds.gen_hsmls:
+            return
+        # TODO: We need to do something like this here but I'm not sure how
+        """
+        if os.path.exists(self.hsml_filename):
+            with open(self.hsml_filename, 'rb') as f:
+                file_hash = struct.unpack('q', f.read(struct.calcsize('q')))[0]
+            if file_hash != self.ds._file_hash:
+                os.remove(self.hsml_filename)
+            else:
+                return
+        """
+        positions = []
+        counts = defaultdict(int)
+        for data_file in data_files:
+            for _, ppos in self._yield_coordinates(
+                    data_file, needed_ptype=self.ds._sph_ptypes[0]):
+                counts[data_file.filename] += ppos.shape[0]
+                positions.append(ppos)
+        if not positions:
+            return
+        offsets = {}
+        offset = 0
+        for fn, count in counts.items():
+            offsets[fn] = offset
+            offset += count
+        positions = np.concatenate(positions)[kdtree.idx]
```

Comment:
Is there *any* chance these will be unitful?  If so we should use `uconcatenate`